doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,50 +1,144 @@
+To specify assignable memory locations
+that are interpreted as pointers by the program, one must use
+`__CPROVER_assigns(ptr)` or `__CPROVER_assigns(__CPROVER_typed_target(ptr))`.
+
+This is to ensure that during call by contract replacement they will be havoced
+as pointers, in a type directed way. For instance:
 
 ```c
-__CPROVER_size_t __CPROVER_object_slice(void *ptr, __CPROVER_size_t size);
+struct circular_buffer_t {
+  int *first;
+  int *last;
+  int *current;
+}
+
+void step(struct circular_buffer_t *buf)
+// correct
+__CPROVER_assigns(__CPROVER_typed_target(buf->current))
+// not correct
+__CPROVER_assigns(__CPROVER_object_upto(&(buf->current), sizeof(buf->current))
+{
+  if(buf->current == buf->last)
+    buf->current = buf->first;
+  else
+    buf->current += 1;
+}
 ```
 
-Caveats and limitations: The slices in question must *not*
-be interpreted as pointers by the program. During call-by-contract replacement, 
-`__CPROVER_havoc_slice(ptr, size)` is used to havoc these targets, 
-and `__CPROVER_havoc_slice` does not support havocing pointers. 
-### Parameters
+### Object slice targets
 
-An _assigns_ clause currently supports simple variable types and their pointers,
-structs, and arrays.  Recursive pointer structures are left to future work, as
-their support would require changes to CBMC's memory model.
+Given a pointer `ptr` pointing into some object `o`,
+`__CPROVER_object_whole(ptr)` specifies that all bytes of the object `o`
+are assignable:
+```c
+__CPROVER_assignable_t __CPROVER_object_whole(void *ptr);
+```
+REMARK: this also includes bytes in the object that are before the address given
+by `ptr`;
+
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr)`
+specifies that the range of bytes starting from the pointer and until the end of
+the object `o` are assignable:
+```c
+__CPROVER_assignable_t __CPROVER_object_from(void *ptr);
+```
 
+```c
+const size_t SIZE = 100;
+
+void init_slice_all(char *arr)
+__CPROVER_assigns(__CPROVER_object_from(arr))
+{
+  for(size_t i=0; i<SIZE; i++)
+    arr[i] = 0;
+}
+```
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_upto(ptr, size)`
+specifies that the range of `size` bytes of `o` starting at `ptr` are assignable:
+The `size` value must such that the range does not exceed the object boundary,
+that is, `__CPROVER_object_size(ptr) - __CPROVER_pointer_offset(ptr) >= size` must hold:
+
+```c
+__CPROVER_assignable_t __CPROVER_object_upto(void *ptr, __CPROVER_size_t size);
+```
+
+```c
+const size_t SIZE = 100;
+
+void init_slice_n(char *arr, size_t n)
+__CPROVER_requires(n <= SIZE)
+__CPROVER_assigns(__CPROVER_object_upto(arr, n))
+{
+  for(size_t i=0; i<n; i++)
+    arr[i] = 0;
+}
+```
+### Function parameters
+
+The memory locations storing function parameters values are considered local
+to the function and are hence always assignable.
```

Comment:
Hmm, yes, but the same is true of any other live object. You'll have to phrase this in terms of the constraints that a contract introduces.